src/stdlib_linalg_iterative_aux.fypp

```diff
@@ -0,0 +1,288 @@
+            end do
+        end do
+    end subroutine
+
+    #:endfor    
+
+    #:for k, t, s in R_KINDS_TYPES
+    module subroutine solve_forward_triangular_csr_${s}$(L,b,x)
+        type(CSR_${s}$_type), intent(in) :: L
+        ${t}$, intent(in) :: b(:)
+        ${t}$, intent(inout) :: x(:)
+        ${t}$ :: aux
+    
+        integer(ilp) :: i, j, m
+        select case (L%storage)
+        case(sparse_full)
+            do i = 1, L%nrows
+                aux = zero_${s}$
+                do m = L%rowptr(i), L%rowptr(i+1)-1
+                    j = L%col(m)
+                    if(j>i) cycle ! skip upper part of the matrix
+                    aux = aux + L%data(m)*x(j)
+                end do
+                x(i) = b(i) - aux
+            end do
+        case(sparse_lower)
+            do i = 1, L%nrows
+                aux = zero_${s}$
+                do m = L%rowptr(i), L%rowptr(i+1)-2
+                    j = L%col(m)
+                    aux = aux + L%data(m)*x(j)
+                end do
+                x(i) = b(i) - aux
+            end do
+        case(sparse_upper) ! treates as lower triangular (thus transpose)
+            do i = 1, L%nrows
+                x(i) = b(i)
+                do m = L%rowptr(i)+1, L%rowptr(i+1)-1
+                    j = L%col(m)
+                    x(j) = x(j) - L%data(m)*x(i)
+                end do
+            end do
+        end select
+    end subroutine
+    
+    module subroutine solve_backward_triangular_csr_${s}$(Lt,b,x)
+        type(CSR_${s}$_type), intent(in) :: Lt
+        ${t}$, intent(in) :: b(:)
+        ${t}$, intent(inout) :: x(:)
+        ${t}$ :: aux
+        ${t}$ :: baux(size(x))
+    
+        integer(ilp) :: i, j, m
+    
+        baux = zero_${s}$
+        select case (Lt%storage)
+        case(sparse_full)
+            do i = Lt%nrows, 1, -1
+                x(i) = b(i) - baux(i)
+                do m = Lt%rowptr(i), Lt%rowptr(i+1)-1
+                    j = Lt%col(m)
+                    if(j<i) cycle ! skip lower part of the matrix
+                    baux(j) = baux(j) + Lt%data(m)*x(i)
+                end do
+            end do
+        case(sparse_lower)
+            do i = Lt%nrows, 1, -1
+                x(i) = b(i) - baux(i)
+                do m = Lt%rowptr(i), Lt%rowptr(i+1)-2
+                    j = Lt%col(m)
+                    baux(j) = baux(j) + Lt%data(m)*x(i)
+                end do
+            end do
+        case(sparse_upper) 
+            do i = Lt%nrows, 1, -1
+                x(i) = b(i)
+                do m = Lt%rowptr(i)+1, Lt%rowptr(i+1)-1
+                    j = Lt%col(m)
+                    x(i) = x(i) - Lt%data(m)*x(j)
+                end do
+            end do
+        end select
+    end subroutine
+
+    #:endfor
+    
+
+    !==============================================================
+    ! Factorization for preconditioners
+    !==============================================================
```

Comment:
these procedures could ne used as stand-alone, outside the context of preconditioning, right?